### DISCUSSION TITLE
Integration/Compatible instructions for OPNsense?

### DISCUSSION THREAD

**Discussion by mscdex** [pgj/freebsd-wifibox#50] [Q&A]
Currently I have an OPNsense installation with the wifibox ports installed. The guest boots and everything seems mostly ok, however I am configuring this for hostapd and am at a loss for what exactly is needed. I've added `interface=wlan0` in wifibox's hostapd.conf, which was enough to get hostapd running in the guest, however `hostapd_cli ping` and other commands on the host result in timeouts.

I'm assuming it's because there is some sort of routing or additional network configuration that needs to happen on the host, but whatever it is it would need to not interfere with OPNsense's system configuration.

Ideally I could just start wifibox and then create a wireless interface from within the OPNsense web UI, but when I attempt that it complains that it couldn't create a wireless device for `wifibox0` in AP mode or something to that effect.

Also, FWIW the Intel 9260 is otherwise working with the current version of wifibox with the most recent wifibox-alpine image on FreeBSD 13.0.

**Comment by pgj** [comment]
The `hostapd_cli` will only work from the host if the control sockets are enabled in `hostapd.conf`.  I am sharing you the configuration file I used for testing, maybe this helps:

```
ctrl_interface=/var/run/hostapd
ctrl_interface_group=0
interface=wlan0

ssid=<ssid>
country_code=<country_code>
channel=11
hw_mode=g

ieee80211n=1
ht_capab=[HT40-][SHORT-GI-20][SHORT-GI-40]

auth_algs=1
wpa=2
wpa_passphrase=<wpa_passphrase>
wpa_key_mgmt=WPA-PSK
rsn_pairwise=CCMP

macaddr_acl=0
```

The `wifibox0` interface should be managed by `wifibox` as it has to be linked to the virtual machine via bridging.  There is no "AP mode", it appears as a virtual Ethernet interface on the host.  Unfortunately, I do not have any experience with OPNsense, so I do not have any understanding on how exactly the integration should work.

**Reply by jodyrowens** [reply]
Did this ever get figured out and working?  I am attempting this as an option to get my wifi card as an AP.  I had been using OpenWRT, but it will not load after the first reboot.  It hangs on passing through the wifi card.

**Reply by pgj** [reply]
With native FreeBSD, I was able to make it work.  Of course, it depends on the wireless card -- if that is not supported sufficiently by `bhyve` or the guest Linux system, it will not work.

**Comment by AphidGit** [comment]
I've been trying to get this to function as well. 
So far, I think the best way is to make the 'WIFI' interface a routed subnet. So it can be a /30 that connects to the alpine linux host. 
Then, on the command line, you use the config files to create a /24 on the alpine host. interfaces.conf, udhdcpd.conf, uds_passthru.conf)
Then you use routing rules to route that /24 to the alpine linux as a gateway. 

For my example, I deliberately picked some different IP addresses to make sure I got all the config files changed; 
Alpine linux (wifibox VM): 10.42.16.2
WIFI interface: 10.42.16.2. 

Set a gateway as 10.42.16.2 on WIFI. Set priority to 255 (so it's not set as default gateway, we don't want that, you'd lose internet connectivity). 
Create a static route; route 10.42.17.0/24 via 10.42.16.2. 

Then, within the alpine linux: 
Set its wlan0 as 10.42.17.1/24. Hand out 10.42.17.20 until 10.42.17.200. 

Stop and restart wifibox, then re-apply settings for the interface within the opnsense GUI to give the freebsd side its IP back (when a network interface is live disconnected, the IP is not re-assigned when it reconnects, so you have to do this whenever you run `wifibox stop` commands.). Just hitting apply in the interface seems to give it an IP

So far, everything seems to be starting, though I can't detect any wifi coming up; no SSID visible on my phone at 5cm distance from my gigabit wireless at least while I detect several dozen SSIDs from all over the neighbourhood. There's no relevant error or log messages either even at info level; my card may just be unsupported, I'm not sure. 

``` 
ifconfig wifibox0
wifibox0: flags=1008843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST,LOWER_UP> metric 0 mtu 1500
        description: WIFI (opt2)
        options=0
        ether 58:9c:fc:10:ff:f0
        inet 10.42.16.1 netmask 0xffffff00 broadcast 10.42.16.255
        id 00:00:00:00:00:00 priority 32768 hellotime 2 fwddelay 15
        maxage 20 holdcnt 6 proto rstp maxaddr 2000 timeout 1200
        root id 00:00:00:00:00:00 priority 32768 ifcost 0 port 0
        member: tap1 flags=143<LEARNING,DISCOVER,AUTOEDGE,AUTOPTP>
                ifmaxaddr 0 port 16 priority 128 path cost 2000000
        groups: bridge
        nd6 options=9<PERFORMNUD,IFDISABLED>
```

I can ping 10.42.16.2, so my VM is working. 

``` 
route get 10.42.17.1

route get 10.42.17.1
   route to: 10.42.17.1
destination: 10.42.17.0
       mask: 255.255.255.0
    gateway: 10.42.16.2
        fib: 0
  interface: wifibox0
      flags: <UP,GATEWAY,DONE,STATIC>
 recvpipe  sendpipe  ssthresh  rtt,msec    mtu        weight    expire
       0         0         0         0      1500         1         0
```

Okay, so the route seems good (it routes it to the VM), but if I ping `10.42.17.1`, no reply. The packets are dropped. It does seem to send out packets though, so it's either the return that's dropped by the firewall on opnsense (it's not, there's a pass-all on 10.42.16.0/**22**, which means the 16, 17, 18, and 19 networks I reserved for this purpose should all be allowed through), or the wlan0 interface within alpine linux just doesn't get created. As it's a bit of a black box, it's not sufficiently logging everything back, I can't quite tell what's going wrong (yet).

The manuals, at least, appear to list the exact model of broadcom chip in my wifi card as supported; Broadcom BCM4366 [brcm]

So I got into the console. Apparently, the alpine linux VM doesn't have a wlan adapter at all. 
Though it can see the wireless card on `lspci` , so there's that. Acting on a hunch, I checked the files in `/lib/firmware/brcm`. Indeed, it contained some files for my adapter; 

``` 
lrwxrwxrwx    1 root     root          22 Jun 14 17:44 brcmfmac4366b-pcie.FreeBSD-BHYVE.bin -> brcmfmac4366b-pcie.bin
-rw-r--r--    1 root     root        1.1M Jun 14 17:44 brcmfmac4366b-pcie.bin
lrwxrwxrwx    1 root     root          22 Jun 14 17:44 brcmfmac4366c-pcie.FreeBSD-BHYVE.bin -> brcmfmac4366c-pcie.bin
-rw-r--r--    1 root     root        1.1M Jun 14 17:44 brcmfmac4366c-pcie.bin
```

So that part's good. Not so much when I checked the `dmesg` though: 

``` 
[    1.016866] wl: loading out-of-tree module taints kernel.
[    1.016872] wl: module license 'MIXED/Proprietary' taints kernel.
[    1.016873] Disabling lock debugging due to kernel taint
[    1.016874] wl: module license taints kernel.
[    1.097789] wl driver 6.30.223.271 (r587334) failed with code 1
```

Seems like it's not running well there either.

**Reply by AphidGit** [reply]
I finally got it all fixed up. Anyway; here's the procedure: 

Compile and install as per https://github.com/pgj/freebsd-wifibox-port 
Choose the 'hostapd' method. 
Choose ONLY the driver you need via the 'flavour'. 

You're going to create _two_ LANs and routing rules as described above. 
For the VM, the routing is simple; route everything to the bsd router. 
For the BSD host, route the wireless network (in my case, 10.42.17.0/24) via the VM's main address (in my case, 10.42.16.2). 

For options, you're going to change the following files: 

udp_passthru.conf : Change the network. 

udhcpd.conf:  Remove the `%%DNS%%` garbage. Change the network. 

iptables: Comment out (place `#` in front) for all lines after and including `*nat` 

ip6tables: IPv6 is left as an exercise to the reader. In this case you can just do the same though.

 interfaces.conf: Change the IP addresses.

forwarding.conf: Change the IP addresses.

bhyve.conf: Set `console=yes` , find out where your PCI card is located (for me it was on PCI-8) and set `passthru=8/0/0` to match. 

Compute a WPA PSK using your ssid and password. 

```
wpa_Passphrase --ssid $mySSID --passphrase $secret
``` 

Note that both SSID and secret have an 8-character minimum. 

hostapd.conf: I used the following, adapted from the dev's example. Note that there is one addiional variable to specify: the ISO country code (2-letter code). You can put any, but you're advised to use your own to not annoy other wireless users. Then put the computed psk in $wpaPsk location as below: 

``` 
ctrl_interface=/var/run/hostapd
ctrl_interface_group=0
interface=wlan0

ssid=geraniums
country_code=$IsoCode
channel=11
hw_mode=g

ieee80211n=1
ht_capab=[HT40-][SHORT-GI-20]

auth_algs=1
wpa=2
wpa_psk=$wpaPsk

wpa_key_mgmt=WPA-PSK
rsn_pairwise=CCMP

macaddr_acl=0

```

Finally, start the wifibox. Login to its console, and run the following: 

``` 
ip addr add $myIP dev wlan0
ip route add $myNetwork dev wlan0
 
service hostapd stop
service hostapd start 

service udhcpd stop 
service udhcpd start
```

This is pretty stupid, but I find that if started normally, it doesn't wait long enough for the wifi driver to load and both hostapd and udhcpd will crash. So they need manual restarts every time. 

Also, it doesn't do the routing table correctly, so if you're not masquerading, the routes are messed up and your wifi clients will be able to send you data but the bsd host can't send any back. Manually adding the route fixes it.  

Also needed every time: Login to the firewall, and go to interfaces -> WIFI (or what you named it). Click apply and accept without making edits. 

Every time the power goes out, you will have to SSH into the firewall, then into the wifibox console, then repeat the above commands, then back to the webinterface and re-apply the interface. It's obnoxious and buggy and runs at about a tenth of the speed that it should but at least you have wifi at all, which is an improvement.

**Reply by pgj** [reply]
It should be possible to pre-configure the `wlan0` on the guest system using the `interfaces.conf` file which would remove the need for touching the console.
